// google-ads-node/build/lib/auth.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const google_auth_library_1 = require("google-auth-library");
class Auth {
    constructor(options) {
        this.options = options;
        this.client = this.buildOAuth2Client(this.options.clientId, this.options.clientSecret);
        this.setRefreshToken(this.options.refreshToken);
    }
    async getAccessToken() {
        if (this.options.accessTokenGetter) {
            const token_from_getter = await this.options.accessTokenGetter(this.options.clientId, this.options.clientSecret, this.options.refreshToken);
            return token_from_getter;
        }
        try {
            const { token } = await this.client.getAccessToken();
            return token;
        }
        catch (err) {
            throw new Error(`Failed to refresh access token, reason=${err.message}`);
        }
    }
    buildOAuth2Client(clientId, clientSecret) {
        return new google_auth_library_1.OAuth2Client({
            clientId,
            clientSecret,
        });
    }
    setRefreshToken(refreshToken) {
        const credentials = {
            refresh_token: refreshToken,
        };
        this.client.setCredentials(credentials);
    }
}
exports.default = Auth;
