// google-ads-node/build/lib/logger.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Logger = void 0;
class Logger {
    constructor(options) {
        this.options = {
            output: options.output || "stderr",
            verbosity: options.verbosity || "info",
            callback: options.callback || undefined,
        };
        this.resetLog();
    }
    setRequestMethod(method) {
        this.request.request.method = method;
    }
    setRequestHeaders(headers) {
        this.request.request.headers = headers;
    }
    setRequestBody(body) {
        this.request.request.body = body;
    }
    setResponseHeaders(headers) {
        this.request.response.headers = headers;
    }
    setResponseBody(body) {
        this.request.response.body = body;
    }
    setResponseStatus(status) {
        this.request.response.status = status;
    }
    setRequestIsMutation() {
        this.request.meta.is_mutation = true;
    }
    setStartTs() {
        this.start_ts = new Date();
    }
    setEndTs() {
        this.end_ts = new Date();
    }
    calculateElapsedMs() {
        if (!this.end_ts || !this.start_ts) {
            return;
        }
        // @ts-ignore This is okay
        const ms = this.end_ts - this.start_ts;
        this.request.meta.elapsed_ms = ms;
    }
    log() {
        let output = "";
        this.calculateElapsedMs();
        if (this.options.verbosity === "warning") {
            output = this.getWarningMessage();
        }
        else if (this.options.verbosity === "debug") {
            output = this.getDebugMessage();
        }
        else {
            output = this.getInfoMessage();
        }
        if (this.options.output === "stdout") {
            process.stdout.write(output);
        }
        if (this.options.output === "stderr") {
            process.stderr.write(output);
        }
        if (this.options.callback) {
            this.options.callback(this.request);
        }
        this.resetLog();
    }
    getDebugMessage() {
        return `\n${JSON.stringify(this.request.request)}\n`;
    }
    getInfoMessage() {
        return `\n${JSON.stringify(this.request)}\n`;
    }
    getWarningMessage() {
        return `\n${JSON.stringify(this.request.response)}\n`;
    }
    resetLog() {
        const default_meta = {
            is_mutation: false,
            elapsed_ms: undefined,
        };
        this.request = {
            request: { method: undefined, headers: undefined, body: undefined },
            response: { headers: undefined, body: undefined, status: undefined },
            meta: default_meta,
        };
        this.start_ts = undefined;
        this.end_ts = undefined;
    }
}
exports.Logger = Logger;
