// google-ads-node/build/lib/mapping.js
"use strict";
/* Autogenerated File! Do Not Edit. */
Object.defineProperty(exports, "__esModule", { value: true });
exports.video = exports.user_location_view = exports.user_list = exports.user_interest = exports.topic_view = exports.topic_constant = exports.shopping_performance_view = exports.shared_set = exports.shared_criterion = exports.search_term_view = exports.remarketing_action = exports.recommendation = exports.call_extension_recommendation = exports.sitelink_extension_recommendation = exports.optimize_ad_rotation_recommendation = exports.search_partners_opt_in_recommendation = exports.keyword_match_type_recommendation = exports.enhanced_cpc_opt_in_recommendation = exports.maximize_conversions_opt_in_recommendation = exports.callout_extension_recommendation = exports.target_cpa_opt_in_recommendation = exports.target_cpa_opt_in_recommendation_option = exports.move_unused_budget_recommendation = exports.text_ad_recommendation = exports.maximize_clicks_opt_in_recommendation = exports.keyword_recommendation = exports.campaign_budget_recommendation = exports.campaign_budget_recommendation_option = exports.recommendation_metrics = exports.recommendation_impact = exports.product_group_view = exports.product_bidding_category_constant = exports.payments_account = exports.parental_status_view = exports.paid_organic_search_term_view = exports.operating_system_version_constant = exports.mutate_job = exports.mutate_job_metadata = exports.mobile_device_constant = exports.mobile_app_category_constant = exports.merchant_center_link = exports.media_video = exports.media_audio = exports.media_bundle = exports.media_image = exports.media_file = exports.managed_placement_view = exports.location_view = exports.language_constant = exports.landing_page_view = exports.label = exports.keyword_view = exports.keyword_plan_negative_keyword = exports.keyword_plan_keyword = exports.keyword_plan_geo_target = exports.keyword_plan_campaign = exports.keyword_plan_ad_group = exports.keyword_plan_forecast_period = exports.keyword_plan = exports.invoice = exports.account_budget_summary = exports.hotel_performance_view = exports.hotel_group_view = exports.group_placement_view = exports.google_ads_field = exports.geographic_view = exports.geo_target_constant = exports.gender_view = exports.feed_placeholder_view = exports.attribute_field_mapping = exports.feed_mapping = exports.feed_item_target = exports.feed_item_validation_error = exports.feed_item_placeholder_policy_info = exports.feed_item_attribute_value = exports.feed_item = exports.feed_attribute_operation = exports.operator = exports.feed_attribute = exports.feed = exports.affiliate_location_feed_data = exports.places_location_feed_data = exports.o_auth_info = exports.extension_feed_item = exports.expanded_landing_page_view = exports.dynamic_search_ads_search_term_view = exports.domain_category = exports.distance_view = exports.display_keyword_view = exports.detail_placement_view = exports.customer_negative_criterion = exports.customer_manager_link = exports.customer_label = exports.customer_feed = exports.customer_extension_setting = exports.customer_client_link = exports.customer_client = exports.remarketing_setting = exports.conversion_tracking_setting = exports.call_reporting_setting = exports.customer = exports.custom_interest_member = exports.custom_interest = exports.currency_constant = exports.conversion_action = exports.value_settings = exports.attribution_model_settings = exports.click_view = exports.change_status = exports.carrier_constant = exports.campaign_shared_set = exports.campaign_label = exports.campaign_feed = exports.campaign_extension_setting = exports.campaign_experiment = exports.campaign_draft = exports.campaign_criterion_simulation = exports.campaign_criterion = exports.campaign_budget = exports.campaign_bid_modifier = exports.campaign_audience_view = exports.campaign = exports.tracking_setting = exports.geo_target_type_setting = exports.selective_optimization = exports.vanity_pharma = exports.shopping_setting = exports.dynamic_search_ads_setting = exports.app_campaign_setting = exports.hotel_setting_info = exports.network_settings = exports.billing_setup = exports.payments_account_info = exports.bidding_strategy = exports.asset = exports.age_range_view = exports.ad_schedule_view = exports.ad_parameter = exports.ad_group_simulation = exports.ad_group_label = exports.ad_group_feed = exports.ad_group_extension_setting = exports.ad_group_criterion_simulation = exports.ad_group_criterion_label = exports.ad_group_criterion = exports.position_estimates = exports.quality_info = exports.ad_group_bid_modifier = exports.ad_group_audience_view = exports.ad_group_ad_label = exports.ad_group_ad_asset_policy_summary = exports.ad_group_ad_asset_view = exports.ad_group_ad_policy_summary = exports.ad_group_ad = exports.ad_group = exports.ad = exports.account_budget_proposal = exports.account_budget = exports.pending_account_budget_proposal = exports.user_list_action_info = exports.basic_user_list_info = exports.logical_user_list_operand_info = exports.user_list_logical_rule_info = exports.logical_user_list_info = exports.rule_based_user_list_info = exports.expression_rule_user_list_info = exports.date_specific_rule_user_list_info = exports.combined_rule_user_list_info = exports.user_list_string_rule_item_info = exports.user_list_number_rule_item_info = exports.user_list_date_rule_item_info = exports.user_list_rule_item_info = exports.user_list_rule_item_group_info = exports.user_list_rule_info = exports.crm_based_user_list_info = exports.similar_user_list_info = exports.url_collection = exports.text_label = exports.target_restriction_operation = exports.target_restriction_operator = exports.target_restriction = exports.targeting_setting = exports.tag_snippet = exports.target_cpa_simulation_point = exports.cpv_bid_simulation_point = exports.cpc_bid_simulation_point = exports.bid_modifier_simulation_point = exports.target_cpa_simulation_point_list = exports.cpv_bid_simulation_point_list = exports.cpc_bid_simulation_point_list = exports.bid_modifier_simulation_point_list = exports.keyword = exports.segments = exports.real_time_bidding_setting = exports.policy_topic_entry = exports.policy_validation_parameter = exports.policy_violation_key = exports.metrics = exports.operand = exports.request_context_operand = exports.function_operand = exports.feed_attribute_operand = exports.constant_operand = exports.matching_function = exports.keyword_plan_historical_metrics = exports.frequency_cap_key = exports.frequency_cap_entry = exports.final_app_url = exports.money = exports.hotel_callout_feed_item = exports.sitelink_feed_item = exports.structured_snippet_feed_item = exports.promotion_feed_item = exports.price_offer = exports.price_feed_item = exports.text_message_feed_item = exports.affiliate_location_feed_item = exports.location_feed_item = exports.callout_feed_item = exports.call_feed_item = exports.app_feed_item = exports.explorer_auto_optimizer_setting = exports.date_range = exports.custom_parameter = exports.criterion_category_locale_availability = exports.criterion_category_channel_availability = exports.criterion_category_availability = exports.location_group_info = exports.custom_intent_info = exports.custom_affinity_info = exports.mobile_device_info = exports.app_payment_model_info = exports.operating_system_version_info = exports.webpage_condition_info = exports.webpage_info = exports.user_interest_info = exports.carrier_info = exports.content_label_info = exports.ip_block_info = exports.language_info = exports.topic_info = exports.address_info = exports.geo_point_info = exports.proximity_info = exports.user_list_info = exports.you_tube_channel_info = exports.you_tube_video_info = exports.parental_status_info = exports.income_range_info = exports.gender_info = exports.age_range_info = exports.ad_schedule_info = exports.interaction_type_info = exports.hotel_check_in_day_info = exports.hotel_length_of_stay_info = exports.hotel_advance_booking_window_info = exports.hotel_date_selection_type_info = exports.unknown_listing_dimension_info = exports.product_type_info = exports.product_item_id_info = exports.product_custom_attribute_info = exports.product_condition_info = exports.product_channel_exclusivity_info = exports.product_channel_info = exports.product_brand_info = exports.product_bidding_category_info = exports.hotel_city_info = exports.hotel_state_info = exports.hotel_country_region_info = exports.hotel_class_info = exports.hotel_id_info = exports.listing_dimension_info = exports.listing_scope_info = exports.listing_group_info = exports.preferred_content_info = exports.device_info = exports.location_info = exports.mobile_application_info = exports.mobile_app_category_info = exports.placement_info = exports.keyword_info = exports.click_location = exports.percent_cpc = exports.target_spend = exports.target_roas = exports.target_impression_share = exports.target_cpm = exports.target_cpa = exports.maximize_conversion_value = exports.maximize_conversions = exports.manual_cpv = exports.manual_cpm = exports.manual_cpc = exports.enhanced_cpc = exports.commission = exports.text_asset = exports.image_dimension = exports.image_asset = exports.media_bundle_asset = exports.youtube_video_asset = exports.display_upload_ad_info = exports.responsive_display_ad_info = exports.legacy_app_install_ad_info = exports.app_engagement_ad_info = exports.app_ad_info = exports.legacy_responsive_display_ad_info = exports.responsive_search_ad_info = exports.video_ad_info = exports.video_true_view_discovery_ad_info = exports.video_outstream_ad_info = exports.video_true_view_in_stream_ad_info = exports.video_non_skippable_in_stream_ad_info = exports.video_bumper_in_stream_ad_info = exports.image_ad_info = exports.product_video = exports.product_image = exports.display_call_to_action = exports.gmail_teaser = exports.gmail_ad_info = exports.shopping_comparison_listing_ad_info = exports.shopping_product_ad_info = exports.shopping_smart_ad_info = exports.hotel_ad_info = exports.expanded_dynamic_search_ad_info = exports.call_only_ad_info = exports.expanded_text_ad_info = exports.text_ad_info = exports.ad_media_bundle_asset = exports.ad_video_asset = exports.ad_image_asset = exports.ad_text_asset = void 0;
/* .google.ads.googleads.v3.common.AdTextAsset */
exports.ad_text_asset = {
    pinned_field: "ServedAssetFieldType",
};
/* .google.ads.googleads.v3.common.AdImageAsset */
exports.ad_image_asset = {};
/* .google.ads.googleads.v3.common.AdVideoAsset */
exports.ad_video_asset = {};
/* .google.ads.googleads.v3.common.AdMediaBundleAsset */
exports.ad_media_bundle_asset = {};
/* .google.ads.googleads.v3.common.TextAdInfo */
exports.text_ad_info = {};
/* .google.ads.googleads.v3.common.ExpandedTextAdInfo */
exports.expanded_text_ad_info = {};
/* .google.ads.googleads.v3.common.CallOnlyAdInfo */
exports.call_only_ad_info = {
    conversion_reporting_state: "CallConversionReportingState",
};
/* .google.ads.googleads.v3.common.ExpandedDynamicSearchAdInfo */
exports.expanded_dynamic_search_ad_info = {};
/* .google.ads.googleads.v3.common.HotelAdInfo */
exports.hotel_ad_info = {};
/* .google.ads.googleads.v3.common.ShoppingSmartAdInfo */
exports.shopping_smart_ad_info = {};
/* .google.ads.googleads.v3.common.ShoppingProductAdInfo */
exports.shopping_product_ad_info = {};
/* .google.ads.googleads.v3.common.ShoppingComparisonListingAdInfo */
exports.shopping_comparison_listing_ad_info = {};
/* .google.ads.googleads.v3.common.GmailAdInfo */
exports.gmail_ad_info = {};
/* .google.ads.googleads.v3.common.GmailTeaser */
exports.gmail_teaser = {};
/* .google.ads.googleads.v3.common.DisplayCallToAction */
exports.display_call_to_action = {};
/* .google.ads.googleads.v3.common.ProductImage */
exports.product_image = {
    display_call_to_action: exports.display_call_to_action,
};
/* .google.ads.googleads.v3.common.ProductVideo */
exports.product_video = {};
/* .google.ads.googleads.v3.common.ImageAdInfo */
exports.image_ad_info = {
    mime_type: "MimeType",
};
/* .google.ads.googleads.v3.common.VideoBumperInStreamAdInfo */
exports.video_bumper_in_stream_ad_info = {};
/* .google.ads.googleads.v3.common.VideoNonSkippableInStreamAdInfo */
exports.video_non_skippable_in_stream_ad_info = {};
/* .google.ads.googleads.v3.common.VideoTrueViewInStreamAdInfo */
exports.video_true_view_in_stream_ad_info = {};
/* .google.ads.googleads.v3.common.VideoOutstreamAdInfo */
exports.video_outstream_ad_info = {};
/* .google.ads.googleads.v3.common.VideoTrueViewDiscoveryAdInfo */
exports.video_true_view_discovery_ad_info = {};
/* .google.ads.googleads.v3.common.VideoAdInfo */
exports.video_ad_info = {
    in_stream: exports.video_true_view_in_stream_ad_info,
    bumper: exports.video_bumper_in_stream_ad_info,
    out_stream: exports.video_outstream_ad_info,
    non_skippable: exports.video_non_skippable_in_stream_ad_info,
    discovery: exports.video_true_view_discovery_ad_info,
};
/* .google.ads.googleads.v3.common.ResponsiveSearchAdInfo */
exports.responsive_search_ad_info = {
    headlines: exports.ad_text_asset,
    descriptions: exports.ad_text_asset,
};
/* .google.ads.googleads.v3.common.LegacyResponsiveDisplayAdInfo */
exports.legacy_responsive_display_ad_info = {
    format_setting: "DisplayAdFormatSetting",
};
/* .google.ads.googleads.v3.common.AppAdInfo */
exports.app_ad_info = {
    mandatory_ad_text: exports.ad_text_asset,
    headlines: exports.ad_text_asset,
    descriptions: exports.ad_text_asset,
    images: exports.ad_image_asset,
    youtube_videos: exports.ad_video_asset,
    html_5_media_bundles: exports.ad_media_bundle_asset,
};
/* .google.ads.googleads.v3.common.AppEngagementAdInfo */
exports.app_engagement_ad_info = {
    headlines: exports.ad_text_asset,
    descriptions: exports.ad_text_asset,
    images: exports.ad_image_asset,
    videos: exports.ad_video_asset,
};
/* .google.ads.googleads.v3.common.LegacyAppInstallAdInfo */
exports.legacy_app_install_ad_info = {
    app_store: "LegacyAppInstallAdAppStore",
};
/* .google.ads.googleads.v3.common.ResponsiveDisplayAdInfo */
exports.responsive_display_ad_info = {
    marketing_images: exports.ad_image_asset,
    square_marketing_images: exports.ad_image_asset,
    logo_images: exports.ad_image_asset,
    square_logo_images: exports.ad_image_asset,
    headlines: exports.ad_text_asset,
    long_headline: exports.ad_text_asset,
    descriptions: exports.ad_text_asset,
    youtube_videos: exports.ad_video_asset,
    format_setting: "DisplayAdFormatSetting",
};
/* .google.ads.googleads.v3.common.DisplayUploadAdInfo */
exports.display_upload_ad_info = {
    display_upload_product_type: "DisplayUploadProductType",
    media_bundle: exports.ad_media_bundle_asset,
};
/* .google.ads.googleads.v3.common.YoutubeVideoAsset */
exports.youtube_video_asset = {};
/* .google.ads.googleads.v3.common.MediaBundleAsset */
exports.media_bundle_asset = {};
/* .google.ads.googleads.v3.common.ImageAsset */
exports.image_asset = {
    mime_type: "MimeType",
};
/* .google.ads.googleads.v3.common.ImageDimension */
exports.image_dimension = {};
/* .google.ads.googleads.v3.common.TextAsset */
exports.text_asset = {};
/* .google.ads.googleads.v3.common.Commission */
exports.commission = {};
/* .google.ads.googleads.v3.common.EnhancedCpc */
exports.enhanced_cpc = {};
/* .google.ads.googleads.v3.common.ManualCpc */
exports.manual_cpc = {};
/* .google.ads.googleads.v3.common.ManualCpm */
exports.manual_cpm = {};
/* .google.ads.googleads.v3.common.ManualCpv */
exports.manual_cpv = {};
/* .google.ads.googleads.v3.common.MaximizeConversions */
exports.maximize_conversions = {};
/* .google.ads.googleads.v3.common.MaximizeConversionValue */
exports.maximize_conversion_value = {};
/* .google.ads.googleads.v3.common.TargetCpa */
exports.target_cpa = {};
/* .google.ads.googleads.v3.common.TargetCpm */
exports.target_cpm = {};
/* .google.ads.googleads.v3.common.TargetImpressionShare */
exports.target_impression_share = {
    location: "TargetImpressionShareLocation",
};
/* .google.ads.googleads.v3.common.TargetRoas */
exports.target_roas = {};
/* .google.ads.googleads.v3.common.TargetSpend */
exports.target_spend = {};
/* .google.ads.googleads.v3.common.PercentCpc */
exports.percent_cpc = {};
/* .google.ads.googleads.v3.common.ClickLocation */
exports.click_location = {};
/* .google.ads.googleads.v3.common.KeywordInfo */
exports.keyword_info = {
    match_type: "KeywordMatchType",
};
/* .google.ads.googleads.v3.common.PlacementInfo */
exports.placement_info = {};
/* .google.ads.googleads.v3.common.MobileAppCategoryInfo */
exports.mobile_app_category_info = {};
/* .google.ads.googleads.v3.common.MobileApplicationInfo */
exports.mobile_application_info = {};
/* .google.ads.googleads.v3.common.LocationInfo */
exports.location_info = {};
/* .google.ads.googleads.v3.common.DeviceInfo */
exports.device_info = {
    type: "Device",
};
/* .google.ads.googleads.v3.common.PreferredContentInfo */
exports.preferred_content_info = {
    type: "PreferredContentType",
};
/* .google.ads.googleads.v3.common.ListingGroupInfo */
exports.listing_group_info = {
    type: "ListingGroupType",
};
/* .google.ads.googleads.v3.common.ListingScopeInfo */
exports.listing_scope_info = {};
/* .google.ads.googleads.v3.common.ListingDimensionInfo */
exports.listing_dimension_info = {};
/* .google.ads.googleads.v3.common.HotelIdInfo */
exports.hotel_id_info = {};
/* .google.ads.googleads.v3.common.HotelClassInfo */
exports.hotel_class_info = {};
/* .google.ads.googleads.v3.common.HotelCountryRegionInfo */
exports.hotel_country_region_info = {};
/* .google.ads.googleads.v3.common.HotelStateInfo */
exports.hotel_state_info = {};
/* .google.ads.googleads.v3.common.HotelCityInfo */
exports.hotel_city_info = {};
/* .google.ads.googleads.v3.common.ProductBiddingCategoryInfo */
exports.product_bidding_category_info = {
    level: "ProductBiddingCategoryLevel",
};
/* .google.ads.googleads.v3.common.ProductBrandInfo */
exports.product_brand_info = {};
/* .google.ads.googleads.v3.common.ProductChannelInfo */
exports.product_channel_info = {
    channel: "ProductChannel",
};
/* .google.ads.googleads.v3.common.ProductChannelExclusivityInfo */
exports.product_channel_exclusivity_info = {
    channel_exclusivity: "ProductChannelExclusivity",
};
/* .google.ads.googleads.v3.common.ProductConditionInfo */
exports.product_condition_info = {
    condition: "ProductCondition",
};
/* .google.ads.googleads.v3.common.ProductCustomAttributeInfo */
exports.product_custom_attribute_info = {
    index: "ProductCustomAttributeIndex",
};
/* .google.ads.googleads.v3.common.ProductItemIdInfo */
exports.product_item_id_info = {};
/* .google.ads.googleads.v3.common.ProductTypeInfo */
exports.product_type_info = {
    level: "ProductTypeLevel",
};
/* .google.ads.googleads.v3.common.UnknownListingDimensionInfo */
exports.unknown_listing_dimension_info = {};
/* .google.ads.googleads.v3.common.HotelDateSelectionTypeInfo */
exports.hotel_date_selection_type_info = {
    type: "HotelDateSelectionType",
};
/* .google.ads.googleads.v3.common.HotelAdvanceBookingWindowInfo */
exports.hotel_advance_booking_window_info = {};
/* .google.ads.googleads.v3.common.HotelLengthOfStayInfo */
exports.hotel_length_of_stay_info = {};
/* .google.ads.googleads.v3.common.HotelCheckInDayInfo */
exports.hotel_check_in_day_info = {
    day_of_week: "DayOfWeek",
};
/* .google.ads.googleads.v3.common.InteractionTypeInfo */
exports.interaction_type_info = {
    type: "InteractionType",
};
/* .google.ads.googleads.v3.common.AdScheduleInfo */
exports.ad_schedule_info = {
    start_minute: "MinuteOfHour",
    end_minute: "MinuteOfHour",
    day_of_week: "DayOfWeek",
};
/* .google.ads.googleads.v3.common.AgeRangeInfo */
exports.age_range_info = {
    type: "AgeRangeType",
};
/* .google.ads.googleads.v3.common.GenderInfo */
exports.gender_info = {
    type: "GenderType",
};
/* .google.ads.googleads.v3.common.IncomeRangeInfo */
exports.income_range_info = {
    type: "IncomeRangeType",
};
/* .google.ads.googleads.v3.common.ParentalStatusInfo */
exports.parental_status_info = {
    type: "ParentalStatusType",
};
/* .google.ads.googleads.v3.common.YouTubeVideoInfo */
exports.you_tube_video_info = {};
/* .google.ads.googleads.v3.common.YouTubeChannelInfo */
exports.you_tube_channel_info = {};
/* .google.ads.googleads.v3.common.UserListInfo */
exports.user_list_info = {};
/* .google.ads.googleads.v3.common.ProximityInfo */
exports.proximity_info = {
    radius_units: "ProximityRadiusUnits",
};
/* .google.ads.googleads.v3.common.GeoPointInfo */
exports.geo_point_info = {};
/* .google.ads.googleads.v3.common.AddressInfo */
exports.address_info = {};
/* .google.ads.googleads.v3.common.TopicInfo */
exports.topic_info = {};
/* .google.ads.googleads.v3.common.LanguageInfo */
exports.language_info = {};
/* .google.ads.googleads.v3.common.IpBlockInfo */
exports.ip_block_info = {};
/* .google.ads.googleads.v3.common.ContentLabelInfo */
exports.content_label_info = {
    type: "ContentLabelType",
};
/* .google.ads.googleads.v3.common.CarrierInfo */
exports.carrier_info = {};
/* .google.ads.googleads.v3.common.UserInterestInfo */
exports.user_interest_info = {};
/* .google.ads.googleads.v3.common.WebpageInfo */
exports.webpage_info = {};
/* .google.ads.googleads.v3.common.WebpageConditionInfo */
exports.webpage_condition_info = {
    operand: "WebpageConditionOperand",
    operator: "WebpageConditionOperator",
};
/* .google.ads.googleads.v3.common.OperatingSystemVersionInfo */
exports.operating_system_version_info = {};
/* .google.ads.googleads.v3.common.AppPaymentModelInfo */
exports.app_payment_model_info = {
    type: "AppPaymentModelType",
};
/* .google.ads.googleads.v3.common.MobileDeviceInfo */
exports.mobile_device_info = {};
/* .google.ads.googleads.v3.common.CustomAffinityInfo */
exports.custom_affinity_info = {};
/* .google.ads.googleads.v3.common.CustomIntentInfo */
exports.custom_intent_info = {};
/* .google.ads.googleads.v3.common.LocationGroupInfo */
exports.location_group_info = {
    radius_units: "LocationGroupRadiusUnits",
};
/* .google.ads.googleads.v3.common.CriterionCategoryAvailability */
exports.criterion_category_availability = {};
/* .google.ads.googleads.v3.common.CriterionCategoryChannelAvailability */
exports.criterion_category_channel_availability = {
    availability_mode: "CriterionCategoryChannelAvailabilityMode",
    advertising_channel_type: "AdvertisingChannelType",
    advertising_channel_sub_type: "AdvertisingChannelSubType",
};
/* .google.ads.googleads.v3.common.CriterionCategoryLocaleAvailability */
exports.criterion_category_locale_availability = {
    availability_mode: "CriterionCategoryLocaleAvailabilityMode",
};
/* .google.ads.googleads.v3.common.CustomParameter */
exports.custom_parameter = {};
/* .google.ads.googleads.v3.common.DateRange */
exports.date_range = {};
/* .google.ads.googleads.v3.common.ExplorerAutoOptimizerSetting */
exports.explorer_auto_optimizer_setting = {};
/* .google.ads.googleads.v3.common.AppFeedItem */
exports.app_feed_item = {
    app_store: "AppStore",
    url_custom_parameters: exports.custom_parameter,
};
/* .google.ads.googleads.v3.common.CallFeedItem */
exports.call_feed_item = {
    call_conversion_reporting_state: "CallConversionReportingState",
};
/* .google.ads.googleads.v3.common.CalloutFeedItem */
exports.callout_feed_item = {};
/* .google.ads.googleads.v3.common.LocationFeedItem */
exports.location_feed_item = {};
/* .google.ads.googleads.v3.common.AffiliateLocationFeedItem */
exports.affiliate_location_feed_item = {};
/* .google.ads.googleads.v3.common.TextMessageFeedItem */
exports.text_message_feed_item = {};
/* .google.ads.googleads.v3.common.PriceFeedItem */
exports.price_feed_item = {
    type: "PriceExtensionType",
    price_qualifier: "PriceExtensionPriceQualifier",
};
/* .google.ads.googleads.v3.common.PriceOffer */
exports.price_offer = {
    unit: "PriceExtensionPriceUnit",
};
/* .google.ads.googleads.v3.common.PromotionFeedItem */
exports.promotion_feed_item = {
    discount_modifier: "PromotionExtensionDiscountModifier",
    occasion: "PromotionExtensionOccasion",
    url_custom_parameters: exports.custom_parameter,
};
/* .google.ads.googleads.v3.common.StructuredSnippetFeedItem */
exports.structured_snippet_feed_item = {};
/* .google.ads.googleads.v3.common.SitelinkFeedItem */
exports.sitelink_feed_item = {
    url_custom_parameters: exports.custom_parameter,
};
/* .google.ads.googleads.v3.common.HotelCalloutFeedItem */
exports.hotel_callout_feed_item = {};
/* .google.ads.googleads.v3.common.Money */
exports.money = {};
/* .google.ads.googleads.v3.common.FinalAppUrl */
exports.final_app_url = {
    os_type: "AppUrlOperatingSystemType",
};
/* .google.ads.googleads.v3.common.FrequencyCapEntry */
exports.frequency_cap_entry = {};
/* .google.ads.googleads.v3.common.FrequencyCapKey */
exports.frequency_cap_key = {
    level: "FrequencyCapLevel",
    event_type: "FrequencyCapEventType",
    time_unit: "FrequencyCapTimeUnit",
};
/* .google.ads.googleads.v3.common.KeywordPlanHistoricalMetrics */
exports.keyword_plan_historical_metrics = {
    competition: "KeywordPlanCompetitionLevel",
};
/* .google.ads.googleads.v3.common.MatchingFunction */
exports.matching_function = {
    operator: "MatchingFunctionOperator",
};
/* .google.ads.googleads.v3.common.Operand.ConstantOperand */
exports.constant_operand = {};
/* .google.ads.googleads.v3.common.Operand.FeedAttributeOperand */
exports.feed_attribute_operand = {};
/* .google.ads.googleads.v3.common.Operand.FunctionOperand */
exports.function_operand = {
    matching_function: exports.matching_function,
};
/* .google.ads.googleads.v3.common.Operand.RequestContextOperand */
exports.request_context_operand = {
    context_type: "MatchingFunctionContextType",
};
/* .google.ads.googleads.v3.common.Operand */
exports.operand = {
    constant_operand: exports.constant_operand,
    feed_attribute_operand: exports.feed_attribute_operand,
    function_operand: exports.function_operand,
    request_context_operand: exports.request_context_operand,
};
/* .google.ads.googleads.v3.common.Metrics */
exports.metrics = {
    historical_creative_quality_score: "QualityScoreBucket",
    historical_landing_page_quality_score: "QualityScoreBucket",
    historical_search_predicted_ctr: "QualityScoreBucket",
    interaction_event_types: "InteractionEventType",
};
/* .google.ads.googleads.v3.common.PolicyViolationKey */
exports.policy_violation_key = {};
/* .google.ads.googleads.v3.common.PolicyValidationParameter */
exports.policy_validation_parameter = {
    exempt_policy_violation_keys: exports.policy_violation_key,
};
/* .google.ads.googleads.v3.common.PolicyTopicEntry */
exports.policy_topic_entry = {
    type: "PolicyTopicEntryType",
};
/* .google.ads.googleads.v3.common.RealTimeBiddingSetting */
exports.real_time_bidding_setting = {};
/* .google.ads.googleads.v3.common.Segments */
exports.segments = {
    ad_network_type: "AdNetworkType",
    click_type: "ClickType",
    conversion_action_category: "ConversionActionCategory",
    conversion_attribution_event_type: "ConversionAttributionEventType",
    conversion_lag_bucket: "ConversionLagBucket",
    conversion_or_adjustment_lag_bucket: "ConversionOrAdjustmentLagBucket",
    day_of_week: "DayOfWeek",
    device: "Device",
    external_conversion_source: "ExternalConversionSource",
    hotel_check_in_day_of_week: "DayOfWeek",
    hotel_date_selection_type: "HotelDateSelectionType",
    hotel_rate_type: "HotelRateType",
    hotel_price_bucket: "HotelPriceBucket",
    month_of_year: "MonthOfYear",
    placeholder_type: "PlaceholderType",
    product_channel: "ProductChannel",
    product_channel_exclusivity: "ProductChannelExclusivity",
    product_condition: "ProductCondition",
    search_engine_results_page_type: "SearchEngineResultsPageType",
    search_term_match_type: "SearchTermMatchType",
    slot: "Slot",
};
/* .google.ads.googleads.v3.common.Keyword */
exports.keyword = {
    info: exports.keyword_info,
};
/* .google.ads.googleads.v3.common.BidModifierSimulationPointList */
exports.bid_modifier_simulation_point_list = {};
/* .google.ads.googleads.v3.common.CpcBidSimulationPointList */
exports.cpc_bid_simulation_point_list = {};
/* .google.ads.googleads.v3.common.CpvBidSimulationPointList */
exports.cpv_bid_simulation_point_list = {};
/* .google.ads.googleads.v3.common.TargetCpaSimulationPointList */
exports.target_cpa_simulation_point_list = {};
/* .google.ads.googleads.v3.common.BidModifierSimulationPoint */
exports.bid_modifier_simulation_point = {};
/* .google.ads.googleads.v3.common.CpcBidSimulationPoint */
exports.cpc_bid_simulation_point = {};
/* .google.ads.googleads.v3.common.CpvBidSimulationPoint */
exports.cpv_bid_simulation_point = {};
/* .google.ads.googleads.v3.common.TargetCpaSimulationPoint */
exports.target_cpa_simulation_point = {};
/* .google.ads.googleads.v3.common.TagSnippet */
exports.tag_snippet = {
    type: "TrackingCodeType",
    page_format: "TrackingCodePageFormat",
};
/* .google.ads.googleads.v3.common.TargetingSetting */
exports.targeting_setting = {};
/* .google.ads.googleads.v3.common.TargetRestriction */
exports.target_restriction = {
    targeting_dimension: "TargetingDimension",
};
/* .google.ads.googleads.v3.common.TargetRestrictionOperation.Operator */
exports.target_restriction_operator = {};
/* .google.ads.googleads.v3.common.TargetRestrictionOperation */
exports.target_restriction_operation = {
    operator: exports.target_restriction_operator,
    value: exports.target_restriction,
};
/* .google.ads.googleads.v3.common.TextLabel */
exports.text_label = {};
/* .google.ads.googleads.v3.common.UrlCollection */
exports.url_collection = {};
/* .google.ads.googleads.v3.common.SimilarUserListInfo */
exports.similar_user_list_info = {};
/* .google.ads.googleads.v3.common.CrmBasedUserListInfo */
exports.crm_based_user_list_info = {
    upload_key_type: "CustomerMatchUploadKeyType",
    data_source_type: "UserListCrmDataSourceType",
};
/* .google.ads.googleads.v3.common.UserListRuleInfo */
exports.user_list_rule_info = {
    rule_type: "UserListRuleType",
};
/* .google.ads.googleads.v3.common.UserListRuleItemGroupInfo */
exports.user_list_rule_item_group_info = {};
/* .google.ads.googleads.v3.common.UserListRuleItemInfo */
exports.user_list_rule_item_info = {};
/* .google.ads.googleads.v3.common.UserListDateRuleItemInfo */
exports.user_list_date_rule_item_info = {
    operator: "UserListDateRuleItemOperator",
};
/* .google.ads.googleads.v3.common.UserListNumberRuleItemInfo */
exports.user_list_number_rule_item_info = {
    operator: "UserListNumberRuleItemOperator",
};
/* .google.ads.googleads.v3.common.UserListStringRuleItemInfo */
exports.user_list_string_rule_item_info = {
    operator: "UserListStringRuleItemOperator",
};
/* .google.ads.googleads.v3.common.CombinedRuleUserListInfo */
exports.combined_rule_user_list_info = {
    left_operand: exports.user_list_rule_info,
    right_operand: exports.user_list_rule_info,
    rule_operator: "UserListCombinedRuleOperator",
};
/* .google.ads.googleads.v3.common.DateSpecificRuleUserListInfo */
exports.date_specific_rule_user_list_info = {
    rule: exports.user_list_rule_info,
};
/* .google.ads.googleads.v3.common.ExpressionRuleUserListInfo */
exports.expression_rule_user_list_info = {
    rule: exports.user_list_rule_info,
};
/* .google.ads.googleads.v3.common.RuleBasedUserListInfo */
exports.rule_based_user_list_info = {
    prepopulation_status: "UserListPrepopulationStatus",
    combined_rule_user_list: exports.combined_rule_user_list_info,
    date_specific_rule_user_list: exports.date_specific_rule_user_list_info,
    expression_rule_user_list: exports.expression_rule_user_list_info,
};
/* .google.ads.googleads.v3.common.LogicalUserListInfo */
exports.logical_user_list_info = {};
/* .google.ads.googleads.v3.common.UserListLogicalRuleInfo */
exports.user_list_logical_rule_info = {
    operator: "UserListLogicalRuleOperator",
};
/* .google.ads.googleads.v3.common.LogicalUserListOperandInfo */
exports.logical_user_list_operand_info = {};
/* .google.ads.googleads.v3.common.BasicUserListInfo */
exports.basic_user_list_info = {};
/* .google.ads.googleads.v3.common.UserListActionInfo */
exports.user_list_action_info = {};
/* .google.ads.googleads.v3.resources.AccountBudget.PendingAccountBudgetProposal */
exports.pending_account_budget_proposal = {
    proposal_type: "AccountBudgetProposalType",
    end_time_type: "TimeType",
    spending_limit_type: "SpendingLimitType",
};
/* .google.ads.googleads.v3.resources.AccountBudget */
exports.account_budget = {
    status: "AccountBudgetStatus",
    pending_proposal: exports.pending_account_budget_proposal,
    proposed_end_time_type: "TimeType",
    approved_end_time_type: "TimeType",
    proposed_spending_limit_type: "SpendingLimitType",
    approved_spending_limit_type: "SpendingLimitType",
    adjusted_spending_limit_type: "SpendingLimitType",
};
/* .google.ads.googleads.v3.resources.AccountBudgetProposal */
exports.account_budget_proposal = {
    proposal_type: "AccountBudgetProposalType",
    status: "AccountBudgetProposalStatus",
    proposed_start_time_type: "TimeType",
    proposed_end_time_type: "TimeType",
    approved_end_time_type: "TimeType",
    proposed_spending_limit_type: "SpendingLimitType",
    approved_spending_limit_type: "SpendingLimitType",
};
/* .google.ads.googleads.v3.resources.Ad */
exports.ad = {
    final_app_urls: exports.final_app_url,
    url_custom_parameters: exports.custom_parameter,
    type: "AdType",
    device_preference: "Device",
    url_collections: exports.url_collection,
    system_managed_resource_source: "SystemManagedResourceSource",
    text_ad: exports.text_ad_info,
    expanded_text_ad: exports.expanded_text_ad_info,
    call_only_ad: exports.call_only_ad_info,
    expanded_dynamic_search_ad: exports.expanded_dynamic_search_ad_info,
    hotel_ad: exports.hotel_ad_info,
    shopping_smart_ad: exports.shopping_smart_ad_info,
    shopping_product_ad: exports.shopping_product_ad_info,
    gmail_ad: exports.gmail_ad_info,
    image_ad: exports.image_ad_info,
    video_ad: exports.video_ad_info,
    responsive_search_ad: exports.responsive_search_ad_info,
    legacy_responsive_display_ad: exports.legacy_responsive_display_ad_info,
    app_ad: exports.app_ad_info,
    legacy_app_install_ad: exports.legacy_app_install_ad_info,
    responsive_display_ad: exports.responsive_display_ad_info,
    display_upload_ad: exports.display_upload_ad_info,
    app_engagement_ad: exports.app_engagement_ad_info,
    shopping_comparison_listing_ad: exports.shopping_comparison_listing_ad_info,
};
/* .google.ads.googleads.v3.resources.AdGroup */
exports.ad_group = {
    status: "AdGroupStatus",
    type: "AdGroupType",
    ad_rotation_mode: "AdGroupAdRotationMode",
    url_custom_parameters: exports.custom_parameter,
    explorer_auto_optimizer_setting: exports.explorer_auto_optimizer_setting,
    display_custom_bid_dimension: "TargetingDimension",
    targeting_setting: exports.targeting_setting,
    effective_target_cpa_source: "BiddingSource",
    effective_target_roas_source: "BiddingSource",
};
/* .google.ads.googleads.v3.resources.AdGroupAd */
exports.ad_group_ad = {
    status: "AdGroupAdStatus",
    ad: exports.ad,
    ad_strength: "AdStrength",
};
/* .google.ads.googleads.v3.resources.AdGroupAdPolicySummary */
exports.ad_group_ad_policy_summary = {
    policy_topic_entries: exports.policy_topic_entry,
    review_status: "PolicyReviewStatus",
    approval_status: "PolicyApprovalStatus",
};
/* .google.ads.googleads.v3.resources.AdGroupAdAssetView */
exports.ad_group_ad_asset_view = {
    field_type: "AssetFieldType",
    performance_label: "AssetPerformanceLabel",
};
/* .google.ads.googleads.v3.resources.AdGroupAdAssetPolicySummary */
exports.ad_group_ad_asset_policy_summary = {
    policy_topic_entries: exports.policy_topic_entry,
    review_status: "PolicyReviewStatus",
    approval_status: "PolicyApprovalStatus",
};
/* .google.ads.googleads.v3.resources.AdGroupAdLabel */
exports.ad_group_ad_label = {};
/* .google.ads.googleads.v3.resources.AdGroupAudienceView */
exports.ad_group_audience_view = {};
/* .google.ads.googleads.v3.resources.AdGroupBidModifier */
exports.ad_group_bid_modifier = {
    bid_modifier_source: "BidModifierSource",
    hotel_date_selection_type: exports.hotel_date_selection_type_info,
    hotel_advance_booking_window: exports.hotel_advance_booking_window_info,
    hotel_length_of_stay: exports.hotel_length_of_stay_info,
    hotel_check_in_day: exports.hotel_check_in_day_info,
    device: exports.device_info,
    preferred_content: exports.preferred_content_info,
};
/* .google.ads.googleads.v3.resources.AdGroupCriterion.QualityInfo */
exports.quality_info = {
    creative_quality_score: "QualityScoreBucket",
    post_click_quality_score: "QualityScoreBucket",
    search_predicted_ctr: "QualityScoreBucket",
};
/* .google.ads.googleads.v3.resources.AdGroupCriterion.PositionEstimates */
exports.position_estimates = {};
/* .google.ads.googleads.v3.resources.AdGroupCriterion */
exports.ad_group_criterion = {
    status: "AdGroupCriterionStatus",
    quality_info: exports.quality_info,
    type: "CriterionType",
    system_serving_status: "CriterionSystemServingStatus",
    approval_status: "AdGroupCriterionApprovalStatus",
    effective_cpc_bid_source: "BiddingSource",
    effective_cpm_bid_source: "BiddingSource",
    effective_cpv_bid_source: "BiddingSource",
    effective_percent_cpc_bid_source: "BiddingSource",
    position_estimates: exports.position_estimates,
    url_custom_parameters: exports.custom_parameter,
    keyword: exports.keyword_info,
    placement: exports.placement_info,
    mobile_app_category: exports.mobile_app_category_info,
    mobile_application: exports.mobile_application_info,
    listing_group: exports.listing_group_info,
    age_range: exports.age_range_info,
    gender: exports.gender_info,
    income_range: exports.income_range_info,
    parental_status: exports.parental_status_info,
    user_list: exports.user_list_info,
    youtube_video: exports.you_tube_video_info,
    youtube_channel: exports.you_tube_channel_info,
    topic: exports.topic_info,
    user_interest: exports.user_interest_info,
    webpage: exports.webpage_info,
    app_payment_model: exports.app_payment_model_info,
    custom_affinity: exports.custom_affinity_info,
    custom_intent: exports.custom_intent_info,
};
/* .google.ads.googleads.v3.resources.AdGroupCriterionLabel */
exports.ad_group_criterion_label = {};
/* .google.ads.googleads.v3.resources.AdGroupCriterionSimulation */
exports.ad_group_criterion_simulation = {
    type: "SimulationType",
    modification_method: "SimulationModificationMethod",
    cpc_bid_point_list: exports.cpc_bid_simulation_point_list,
};
/* .google.ads.googleads.v3.resources.AdGroupExtensionSetting */
exports.ad_group_extension_setting = {
    extension_type: "ExtensionType",
    device: "ExtensionSettingDevice",
};
/* .google.ads.googleads.v3.resources.AdGroupFeed */
exports.ad_group_feed = {
    placeholder_types: "PlaceholderType",
    matching_function: exports.matching_function,
    status: "FeedLinkStatus",
};
/* .google.ads.googleads.v3.resources.AdGroupLabel */
exports.ad_group_label = {};
/* .google.ads.googleads.v3.resources.AdGroupSimulation */
exports.ad_group_simulation = {
    type: "SimulationType",
    modification_method: "SimulationModificationMethod",
    cpc_bid_point_list: exports.cpc_bid_simulation_point_list,
    cpv_bid_point_list: exports.cpv_bid_simulation_point_list,
    target_cpa_point_list: exports.target_cpa_simulation_point_list,
};
/* .google.ads.googleads.v3.resources.AdParameter */
exports.ad_parameter = {};
/* .google.ads.googleads.v3.resources.AdScheduleView */
exports.ad_schedule_view = {};
/* .google.ads.googleads.v3.resources.AgeRangeView */
exports.age_range_view = {};
/* .google.ads.googleads.v3.resources.Asset */
exports.asset = {
    type: "AssetType",
    youtube_video_asset: exports.youtube_video_asset,
    media_bundle_asset: exports.media_bundle_asset,
    image_asset: exports.image_asset,
    text_asset: exports.text_asset,
};
/* .google.ads.googleads.v3.resources.BiddingStrategy */
exports.bidding_strategy = {
    status: "BiddingStrategyStatus",
    type: "BiddingStrategyType",
    enhanced_cpc: exports.enhanced_cpc,
    target_cpa: exports.target_cpa,
    target_impression_share: exports.target_impression_share,
    target_roas: exports.target_roas,
    target_spend: exports.target_spend,
};
/* .google.ads.googleads.v3.resources.BillingSetup.PaymentsAccountInfo */
exports.payments_account_info = {};
/* .google.ads.googleads.v3.resources.BillingSetup */
exports.billing_setup = {
    status: "BillingSetupStatus",
    payments_account_info: exports.payments_account_info,
    start_time_type: "TimeType",
    end_time_type: "TimeType",
};
/* .google.ads.googleads.v3.resources.Campaign.NetworkSettings */
exports.network_settings = {};
/* .google.ads.googleads.v3.resources.Campaign.HotelSettingInfo */
exports.hotel_setting_info = {};
/* .google.ads.googleads.v3.resources.Campaign.AppCampaignSetting */
exports.app_campaign_setting = {
    bidding_strategy_goal_type: "AppCampaignBiddingStrategyGoalType",
    app_store: "AppCampaignAppStore",
};
/* .google.ads.googleads.v3.resources.Campaign.DynamicSearchAdsSetting */
exports.dynamic_search_ads_setting = {};
/* .google.ads.googleads.v3.resources.Campaign.ShoppingSetting */
exports.shopping_setting = {};
/* .google.ads.googleads.v3.resources.Campaign.VanityPharma */
exports.vanity_pharma = {
    vanity_pharma_display_url_mode: "VanityPharmaDisplayUrlMode",
    vanity_pharma_text: "VanityPharmaText",
};
/* .google.ads.googleads.v3.resources.Campaign.SelectiveOptimization */
exports.selective_optimization = {};
/* .google.ads.googleads.v3.resources.Campaign.GeoTargetTypeSetting */
exports.geo_target_type_setting = {
    positive_geo_target_type: "PositiveGeoTargetType",
    negative_geo_target_type: "NegativeGeoTargetType",
};
/* .google.ads.googleads.v3.resources.Campaign.TrackingSetting */
exports.tracking_setting = {};
/* .google.ads.googleads.v3.resources.Campaign */
exports.campaign = {
    status: "CampaignStatus",
    serving_status: "CampaignServingStatus",
    ad_serving_optimization_status: "AdServingOptimizationStatus",
    advertising_channel_type: "AdvertisingChannelType",
    advertising_channel_sub_type: "AdvertisingChannelSubType",
    url_custom_parameters: exports.custom_parameter,
    real_time_bidding_setting: exports.real_time_bidding_setting,
    network_settings: exports.network_settings,
    hotel_setting: exports.hotel_setting_info,
    dynamic_search_ads_setting: exports.dynamic_search_ads_setting,
    shopping_setting: exports.shopping_setting,
    targeting_setting: exports.targeting_setting,
    geo_target_type_setting: exports.geo_target_type_setting,
    app_campaign_setting: exports.app_campaign_setting,
    experiment_type: "CampaignExperimentType",
    bidding_strategy_type: "BiddingStrategyType",
    frequency_caps: exports.frequency_cap_entry,
    video_brand_safety_suitability: "BrandSafetySuitability",
    vanity_pharma: exports.vanity_pharma,
    selective_optimization: exports.selective_optimization,
    tracking_setting: exports.tracking_setting,
    payment_mode: "PaymentMode",
    commission: exports.commission,
    manual_cpc: exports.manual_cpc,
    manual_cpm: exports.manual_cpm,
    manual_cpv: exports.manual_cpv,
    maximize_conversions: exports.maximize_conversions,
    maximize_conversion_value: exports.maximize_conversion_value,
    target_cpa: exports.target_cpa,
    target_impression_share: exports.target_impression_share,
    target_roas: exports.target_roas,
    target_spend: exports.target_spend,
    percent_cpc: exports.percent_cpc,
    target_cpm: exports.target_cpm,
};
/* .google.ads.googleads.v3.resources.CampaignAudienceView */
exports.campaign_audience_view = {};
/* .google.ads.googleads.v3.resources.CampaignBidModifier */
exports.campaign_bid_modifier = {
    interaction_type: exports.interaction_type_info,
};
/* .google.ads.googleads.v3.resources.CampaignBudget */
exports.campaign_budget = {
    status: "BudgetStatus",
    delivery_method: "BudgetDeliveryMethod",
    period: "BudgetPeriod",
    type: "BudgetType",
};
/* .google.ads.googleads.v3.resources.CampaignCriterion */
exports.campaign_criterion = {
    type: "CriterionType",
    status: "CampaignCriterionStatus",
    keyword: exports.keyword_info,
    placement: exports.placement_info,
    mobile_app_category: exports.mobile_app_category_info,
    mobile_application: exports.mobile_application_info,
    location: exports.location_info,
    device: exports.device_info,
    ad_schedule: exports.ad_schedule_info,
    age_range: exports.age_range_info,
    gender: exports.gender_info,
    income_range: exports.income_range_info,
    parental_status: exports.parental_status_info,
    user_list: exports.user_list_info,
    youtube_video: exports.you_tube_video_info,
    youtube_channel: exports.you_tube_channel_info,
    proximity: exports.proximity_info,
    topic: exports.topic_info,
    listing_scope: exports.listing_scope_info,
    language: exports.language_info,
    ip_block: exports.ip_block_info,
    content_label: exports.content_label_info,
    carrier: exports.carrier_info,
    user_interest: exports.user_interest_info,
    webpage: exports.webpage_info,
    operating_system_version: exports.operating_system_version_info,
    mobile_device: exports.mobile_device_info,
    location_group: exports.location_group_info,
    custom_affinity: exports.custom_affinity_info,
};
/* .google.ads.googleads.v3.resources.CampaignCriterionSimulation */
exports.campaign_criterion_simulation = {
    type: "SimulationType",
    modification_method: "SimulationModificationMethod",
    bid_modifier_point_list: exports.bid_modifier_simulation_point_list,
};
/* .google.ads.googleads.v3.resources.CampaignDraft */
exports.campaign_draft = {
    status: "CampaignDraftStatus",
};
/* .google.ads.googleads.v3.resources.CampaignExperiment */
exports.campaign_experiment = {
    traffic_split_type: "CampaignExperimentTrafficSplitType",
    status: "CampaignExperimentStatus",
};
/* .google.ads.googleads.v3.resources.CampaignExtensionSetting */
exports.campaign_extension_setting = {
    extension_type: "ExtensionType",
    device: "ExtensionSettingDevice",
};
/* .google.ads.googleads.v3.resources.CampaignFeed */
exports.campaign_feed = {
    placeholder_types: "PlaceholderType",
    matching_function: exports.matching_function,
    status: "FeedLinkStatus",
};
/* .google.ads.googleads.v3.resources.CampaignLabel */
exports.campaign_label = {};
/* .google.ads.googleads.v3.resources.CampaignSharedSet */
exports.campaign_shared_set = {
    status: "CampaignSharedSetStatus",
};
/* .google.ads.googleads.v3.resources.CarrierConstant */
exports.carrier_constant = {};
/* .google.ads.googleads.v3.resources.ChangeStatus */
exports.change_status = {
    resource_type: "ChangeStatusResourceType",
    resource_status: "ChangeStatusOperation",
};
/* .google.ads.googleads.v3.resources.ClickView */
exports.click_view = {
    area_of_interest: exports.click_location,
    location_of_presence: exports.click_location,
};
/* .google.ads.googleads.v3.resources.ConversionAction.AttributionModelSettings */
exports.attribution_model_settings = {
    attribution_model: "AttributionModel",
    data_driven_model_status: "DataDrivenModelStatus",
};
/* .google.ads.googleads.v3.resources.ConversionAction.ValueSettings */
exports.value_settings = {};
/* .google.ads.googleads.v3.resources.ConversionAction */
exports.conversion_action = {
    status: "ConversionActionStatus",
    type: "ConversionActionType",
    category: "ConversionActionCategory",
    value_settings: exports.value_settings,
    counting_type: "ConversionActionCountingType",
    attribution_model_settings: exports.attribution_model_settings,
    tag_snippets: exports.tag_snippet,
};
/* .google.ads.googleads.v3.resources.CurrencyConstant */
exports.currency_constant = {};
/* .google.ads.googleads.v3.resources.CustomInterest */
exports.custom_interest = {
    status: "CustomInterestStatus",
    type: "CustomInterestType",
};
/* .google.ads.googleads.v3.resources.CustomInterestMember */
exports.custom_interest_member = {
    member_type: "CustomInterestMemberType",
};
/* .google.ads.googleads.v3.resources.Customer */
exports.customer = {
    pay_per_conversion_eligibility_failure_reasons: "CustomerPayPerConversionEligibilityFailureReason",
};
/* .google.ads.googleads.v3.resources.CallReportingSetting */
exports.call_reporting_setting = {};
/* .google.ads.googleads.v3.resources.ConversionTrackingSetting */
exports.conversion_tracking_setting = {};
/* .google.ads.googleads.v3.resources.RemarketingSetting */
exports.remarketing_setting = {};
/* .google.ads.googleads.v3.resources.CustomerClient */
exports.customer_client = {};
/* .google.ads.googleads.v3.resources.CustomerClientLink */
exports.customer_client_link = {
    status: "ManagerLinkStatus",
};
/* .google.ads.googleads.v3.resources.CustomerExtensionSetting */
exports.customer_extension_setting = {
    extension_type: "ExtensionType",
    device: "ExtensionSettingDevice",
};
/* .google.ads.googleads.v3.resources.CustomerFeed */
exports.customer_feed = {
    placeholder_types: "PlaceholderType",
    matching_function: exports.matching_function,
    status: "FeedLinkStatus",
};
/* .google.ads.googleads.v3.resources.CustomerLabel */
exports.customer_label = {};
/* .google.ads.googleads.v3.resources.CustomerManagerLink */
exports.customer_manager_link = {
    status: "ManagerLinkStatus",
};
/* .google.ads.googleads.v3.resources.CustomerNegativeCriterion */
exports.customer_negative_criterion = {
    type: "CriterionType",
    content_label: exports.content_label_info,
    mobile_application: exports.mobile_application_info,
    mobile_app_category: exports.mobile_app_category_info,
    placement: exports.placement_info,
    youtube_video: exports.you_tube_video_info,
    youtube_channel: exports.you_tube_channel_info,
};
/* .google.ads.googleads.v3.resources.DetailPlacementView */
exports.detail_placement_view = {
    placement_type: "PlacementType",
};
/* .google.ads.googleads.v3.resources.DisplayKeywordView */
exports.display_keyword_view = {};
/* .google.ads.googleads.v3.resources.DistanceView */
exports.distance_view = {
    distance_bucket: "DistanceBucket",
};
/* .google.ads.googleads.v3.resources.DomainCategory */
exports.domain_category = {};
/* .google.ads.googleads.v3.resources.DynamicSearchAdsSearchTermView */
exports.dynamic_search_ads_search_term_view = {};
/* .google.ads.googleads.v3.resources.ExpandedLandingPageView */
exports.expanded_landing_page_view = {};
/* .google.ads.googleads.v3.resources.ExtensionFeedItem */
exports.extension_feed_item = {
    extension_type: "ExtensionType",
    ad_schedules: exports.ad_schedule_info,
    device: "FeedItemTargetDevice",
    targeted_keyword: exports.keyword_info,
    status: "FeedItemStatus",
    sitelink_feed_item: exports.sitelink_feed_item,
    structured_snippet_feed_item: exports.structured_snippet_feed_item,
    app_feed_item: exports.app_feed_item,
    call_feed_item: exports.call_feed_item,
    callout_feed_item: exports.callout_feed_item,
    text_message_feed_item: exports.text_message_feed_item,
    price_feed_item: exports.price_feed_item,
    promotion_feed_item: exports.promotion_feed_item,
    location_feed_item: exports.location_feed_item,
    affiliate_location_feed_item: exports.affiliate_location_feed_item,
    hotel_callout_feed_item: exports.hotel_callout_feed_item,
};
/* .google.ads.googleads.v3.resources.Feed.PlacesLocationFeedData.OAuthInfo */
exports.o_auth_info = {};
/* .google.ads.googleads.v3.resources.Feed.PlacesLocationFeedData */
exports.places_location_feed_data = {
    oauth_info: exports.o_auth_info,
};
/* .google.ads.googleads.v3.resources.Feed.AffiliateLocationFeedData */
exports.affiliate_location_feed_data = {
    relationship_type: "AffiliateLocationFeedRelationshipType",
};
/* .google.ads.googleads.v3.resources.Feed */
exports.feed = {
    origin: "FeedOrigin",
    status: "FeedStatus",
    places_location_feed_data: exports.places_location_feed_data,
    affiliate_location_feed_data: exports.affiliate_location_feed_data,
};
/* .google.ads.googleads.v3.resources.FeedAttribute */
exports.feed_attribute = {
    type: "FeedAttributeType",
};
/* .google.ads.googleads.v3.resources.FeedAttributeOperation.Operator */
exports.operator = {};
/* .google.ads.googleads.v3.resources.FeedAttributeOperation */
exports.feed_attribute_operation = {
    operator: exports.operator,
    value: exports.feed_attribute,
};
/* .google.ads.googleads.v3.resources.FeedItem */
exports.feed_item = {
    geo_targeting_restriction: "GeoTargetingRestriction",
    url_custom_parameters: exports.custom_parameter,
    status: "FeedItemStatus",
};
/* .google.ads.googleads.v3.resources.FeedItemAttributeValue */
exports.feed_item_attribute_value = {
    price_value: exports.money,
};
/* .google.ads.googleads.v3.resources.FeedItemPlaceholderPolicyInfo */
exports.feed_item_placeholder_policy_info = {
    placeholder_type_enum: "PlaceholderType",
    review_status: "PolicyReviewStatus",
    approval_status: "PolicyApprovalStatus",
    policy_topic_entries: exports.policy_topic_entry,
    validation_status: "FeedItemValidationStatus",
    quality_approval_status: "FeedItemQualityApprovalStatus",
    quality_disapproval_reasons: "FeedItemQualityDisapprovalReason",
};
/* .google.ads.googleads.v3.errors.FeedItemValidationErrorEnum.FeedItemValidationError */
exports.feed_item_validation_error = {};
/* .google.ads.googleads.v3.resources.FeedItemTarget */
exports.feed_item_target = {
    feed_item_target_type: "FeedItemTargetType",
    status: "FeedItemTargetStatus",
    keyword: exports.keyword_info,
    device: "FeedItemTargetDevice",
    ad_schedule: exports.ad_schedule_info,
};
/* .google.ads.googleads.v3.resources.FeedMapping */
exports.feed_mapping = {
    status: "FeedMappingStatus",
    placeholder_type: "PlaceholderType",
    criterion_type: "FeedMappingCriterionType",
};
/* .google.ads.googleads.v3.resources.AttributeFieldMapping */
exports.attribute_field_mapping = {
    sitelink_field: "SitelinkPlaceholderField",
    call_field: "CallPlaceholderField",
    app_field: "AppPlaceholderField",
    location_field: "LocationPlaceholderField",
    affiliate_location_field: "AffiliateLocationPlaceholderField",
    callout_field: "CalloutPlaceholderField",
    structured_snippet_field: "StructuredSnippetPlaceholderField",
    message_field: "MessagePlaceholderField",
    price_field: "PricePlaceholderField",
    promotion_field: "PromotionPlaceholderField",
    ad_customizer_field: "AdCustomizerPlaceholderField",
    dsa_page_feed_field: "DsaPageFeedCriterionField",
    location_extension_targeting_field: "LocationExtensionTargetingCriterionField",
    education_field: "EducationPlaceholderField",
    flight_field: "FlightPlaceholderField",
    custom_field: "CustomPlaceholderField",
    hotel_field: "HotelPlaceholderField",
    real_estate_field: "RealEstatePlaceholderField",
    travel_field: "TravelPlaceholderField",
    local_field: "LocalPlaceholderField",
    job_field: "JobPlaceholderField",
};
/* .google.ads.googleads.v3.resources.FeedPlaceholderView */
exports.feed_placeholder_view = {
    placeholder_type: "PlaceholderType",
};
/* .google.ads.googleads.v3.resources.GenderView */
exports.gender_view = {};
/* .google.ads.googleads.v3.resources.GeoTargetConstant */
exports.geo_target_constant = {
    status: "GeoTargetConstantStatus",
};
/* .google.ads.googleads.v3.resources.GeographicView */
exports.geographic_view = {
    location_type: "GeoTargetingType",
};
/* .google.ads.googleads.v3.resources.GoogleAdsField */
exports.google_ads_field = {
    category: "GoogleAdsFieldCategory",
    data_type: "GoogleAdsFieldDataType",
};
/* .google.ads.googleads.v3.resources.GroupPlacementView */
exports.group_placement_view = {
    placement_type: "PlacementType",
};
/* .google.ads.googleads.v3.resources.HotelGroupView */
exports.hotel_group_view = {};
/* .google.ads.googleads.v3.resources.HotelPerformanceView */
exports.hotel_performance_view = {};
/* .google.ads.googleads.v3.resources.Invoice.AccountBudgetSummary */
exports.account_budget_summary = {
    billable_activity_date_range: exports.date_range,
};
/* .google.ads.googleads.v3.resources.Invoice */
exports.invoice = {
    type: "InvoiceType",
    service_date_range: exports.date_range,
    account_budget_summaries: exports.account_budget_summary,
};
/* .google.ads.googleads.v3.resources.KeywordPlan */
exports.keyword_plan = {};
/* .google.ads.googleads.v3.resources.KeywordPlanForecastPeriod */
exports.keyword_plan_forecast_period = {
    date_interval: "KeywordPlanForecastInterval",
    date_range: exports.date_range,
};
/* .google.ads.googleads.v3.resources.KeywordPlanAdGroup */
exports.keyword_plan_ad_group = {};
/* .google.ads.googleads.v3.resources.KeywordPlanCampaign */
exports.keyword_plan_campaign = {
    keyword_plan_network: "KeywordPlanNetwork",
};
/* .google.ads.googleads.v3.resources.KeywordPlanGeoTarget */
exports.keyword_plan_geo_target = {};
/* .google.ads.googleads.v3.resources.KeywordPlanKeyword */
exports.keyword_plan_keyword = {
    match_type: "KeywordMatchType",
};
/* .google.ads.googleads.v3.resources.KeywordPlanNegativeKeyword */
exports.keyword_plan_negative_keyword = {
    match_type: "KeywordMatchType",
};
/* .google.ads.googleads.v3.resources.KeywordView */
exports.keyword_view = {};
/* .google.protobuf.FieldDescriptorProto.Label */
exports.label = {};
/* .google.ads.googleads.v3.resources.LandingPageView */
exports.landing_page_view = {};
/* .google.ads.googleads.v3.resources.LanguageConstant */
exports.language_constant = {};
/* .google.ads.googleads.v3.resources.LocationView */
exports.location_view = {};
/* .google.ads.googleads.v3.resources.ManagedPlacementView */
exports.managed_placement_view = {};
/* .google.ads.googleads.v3.resources.MediaFile */
exports.media_file = {
    type: "MediaType",
    mime_type: "MimeType",
};
/* .google.ads.googleads.v3.resources.MediaImage */
exports.media_image = {};
/* .google.ads.googleads.v3.resources.MediaBundle */
exports.media_bundle = {};
/* .google.ads.googleads.v3.resources.MediaAudio */
exports.media_audio = {};
/* .google.ads.googleads.v3.resources.MediaVideo */
exports.media_video = {};
/* .google.ads.googleads.v3.resources.MerchantCenterLink */
exports.merchant_center_link = {
    status: "MerchantCenterLinkStatus",
};
/* .google.ads.googleads.v3.resources.MobileAppCategoryConstant */
exports.mobile_app_category_constant = {};
/* .google.ads.googleads.v3.resources.MobileDeviceConstant */
exports.mobile_device_constant = {
    type: "MobileDeviceType",
};
/* .google.ads.googleads.v3.resources.MutateJob.MutateJobMetadata */
exports.mutate_job_metadata = {};
/* .google.ads.googleads.v3.resources.MutateJob */
exports.mutate_job = {
    metadata: exports.mutate_job_metadata,
    status: "MutateJobStatus",
};
/* .google.ads.googleads.v3.resources.OperatingSystemVersionConstant */
exports.operating_system_version_constant = {
    operator_type: "OperatingSystemVersionOperatorType",
};
/* .google.ads.googleads.v3.resources.PaidOrganicSearchTermView */
exports.paid_organic_search_term_view = {};
/* .google.ads.googleads.v3.resources.ParentalStatusView */
exports.parental_status_view = {};
/* .google.ads.googleads.v3.resources.PaymentsAccount */
exports.payments_account = {};
/* .google.ads.googleads.v3.resources.ProductBiddingCategoryConstant */
exports.product_bidding_category_constant = {
    level: "ProductBiddingCategoryLevel",
    status: "ProductBiddingCategoryStatus",
};
/* .google.ads.googleads.v3.resources.ProductGroupView */
exports.product_group_view = {};
/* .google.ads.googleads.v3.resources.Recommendation.RecommendationImpact */
exports.recommendation_impact = {};
/* .google.ads.googleads.v3.resources.Recommendation.RecommendationMetrics */
exports.recommendation_metrics = {};
/* .google.ads.googleads.v3.resources.Recommendation.CampaignBudgetRecommendation.CampaignBudgetRecommendationOption */
exports.campaign_budget_recommendation_option = {
    impact: exports.recommendation_impact,
};
/* .google.ads.googleads.v3.resources.Recommendation.CampaignBudgetRecommendation */
exports.campaign_budget_recommendation = {
    budget_options: exports.campaign_budget_recommendation_option,
};
/* .google.ads.googleads.v3.resources.Recommendation.KeywordRecommendation */
exports.keyword_recommendation = {
    keyword: exports.keyword_info,
};
/* .google.ads.googleads.v3.resources.Recommendation.MaximizeClicksOptInRecommendation */
exports.maximize_clicks_opt_in_recommendation = {};
/* .google.ads.googleads.v3.resources.Recommendation.TextAdRecommendation */
exports.text_ad_recommendation = {
    ad: exports.ad,
};
/* .google.ads.googleads.v3.resources.Recommendation.MoveUnusedBudgetRecommendation */
exports.move_unused_budget_recommendation = {
    budget_recommendation: exports.campaign_budget_recommendation,
};
/* .google.ads.googleads.v3.resources.Recommendation.TargetCpaOptInRecommendation.TargetCpaOptInRecommendationOption */
exports.target_cpa_opt_in_recommendation_option = {
    goal: "TargetCpaOptInRecommendationGoal",
    impact: exports.recommendation_impact,
};
/* .google.ads.googleads.v3.resources.Recommendation.TargetCpaOptInRecommendation */
exports.target_cpa_opt_in_recommendation = {
    options: exports.target_cpa_opt_in_recommendation_option,
};
/* .google.ads.googleads.v3.resources.Recommendation.CalloutExtensionRecommendation */
exports.callout_extension_recommendation = {
    recommended_extensions: exports.callout_feed_item,
};
/* .google.ads.googleads.v3.resources.Recommendation.MaximizeConversionsOptInRecommendation */
exports.maximize_conversions_opt_in_recommendation = {};
/* .google.ads.googleads.v3.resources.Recommendation.EnhancedCpcOptInRecommendation */
exports.enhanced_cpc_opt_in_recommendation = {};
/* .google.ads.googleads.v3.resources.Recommendation.KeywordMatchTypeRecommendation */
exports.keyword_match_type_recommendation = {
    keyword: exports.keyword_info,
    recommended_match_type: "KeywordMatchType",
};
/* .google.ads.googleads.v3.resources.Recommendation.SearchPartnersOptInRecommendation */
exports.search_partners_opt_in_recommendation = {};
/* .google.ads.googleads.v3.resources.Recommendation.OptimizeAdRotationRecommendation */
exports.optimize_ad_rotation_recommendation = {};
/* .google.ads.googleads.v3.resources.Recommendation.SitelinkExtensionRecommendation */
exports.sitelink_extension_recommendation = {
    recommended_extensions: exports.sitelink_feed_item,
};
/* .google.ads.googleads.v3.resources.Recommendation.CallExtensionRecommendation */
exports.call_extension_recommendation = {
    recommended_extensions: exports.call_feed_item,
};
/* .google.ads.googleads.v3.resources.Recommendation */
exports.recommendation = {
    type: "RecommendationType",
    impact: exports.recommendation_impact,
    campaign_budget_recommendation: exports.campaign_budget_recommendation,
    keyword_recommendation: exports.keyword_recommendation,
    text_ad_recommendation: exports.text_ad_recommendation,
    target_cpa_opt_in_recommendation: exports.target_cpa_opt_in_recommendation,
    maximize_conversions_opt_in_recommendation: exports.maximize_conversions_opt_in_recommendation,
    enhanced_cpc_opt_in_recommendation: exports.enhanced_cpc_opt_in_recommendation,
    search_partners_opt_in_recommendation: exports.search_partners_opt_in_recommendation,
    maximize_clicks_opt_in_recommendation: exports.maximize_clicks_opt_in_recommendation,
    optimize_ad_rotation_recommendation: exports.optimize_ad_rotation_recommendation,
    callout_extension_recommendation: exports.callout_extension_recommendation,
    sitelink_extension_recommendation: exports.sitelink_extension_recommendation,
    call_extension_recommendation: exports.call_extension_recommendation,
    keyword_match_type_recommendation: exports.keyword_match_type_recommendation,
    move_unused_budget_recommendation: exports.move_unused_budget_recommendation,
};
/* .google.ads.googleads.v3.resources.RemarketingAction */
exports.remarketing_action = {
    tag_snippets: exports.tag_snippet,
};
/* .google.ads.googleads.v3.resources.SearchTermView */
exports.search_term_view = {
    status: "SearchTermTargetingStatus",
};
/* .google.ads.googleads.v3.resources.SharedCriterion */
exports.shared_criterion = {
    type: "CriterionType",
    keyword: exports.keyword_info,
    youtube_video: exports.you_tube_video_info,
    youtube_channel: exports.you_tube_channel_info,
    placement: exports.placement_info,
    mobile_app_category: exports.mobile_app_category_info,
    mobile_application: exports.mobile_application_info,
};
/* .google.ads.googleads.v3.resources.SharedSet */
exports.shared_set = {
    type: "SharedSetType",
    status: "SharedSetStatus",
};
/* .google.ads.googleads.v3.resources.ShoppingPerformanceView */
exports.shopping_performance_view = {};
/* .google.ads.googleads.v3.resources.TopicConstant */
exports.topic_constant = {};
/* .google.ads.googleads.v3.resources.TopicView */
exports.topic_view = {};
/* .google.ads.googleads.v3.resources.UserInterest */
exports.user_interest = {
    taxonomy_type: "UserInterestTaxonomyType",
    availabilities: exports.criterion_category_availability,
};
/* .google.ads.googleads.v3.resources.UserList */
exports.user_list = {
    membership_status: "UserListMembershipStatus",
    size_range_for_display: "UserListSizeRange",
    size_range_for_search: "UserListSizeRange",
    type: "UserListType",
    closing_reason: "UserListClosingReason",
    access_reason: "AccessReason",
    account_user_list_status: "UserListAccessStatus",
    crm_based_user_list: exports.crm_based_user_list_info,
    similar_user_list: exports.similar_user_list_info,
    rule_based_user_list: exports.rule_based_user_list_info,
    logical_user_list: exports.logical_user_list_info,
    basic_user_list: exports.basic_user_list_info,
};
/* .google.ads.googleads.v3.resources.UserLocationView */
exports.user_location_view = {};
/* .google.ads.googleads.v3.resources.Video */
exports.video = {};
