// google-ads-api/build/services/customer.js
"use strict";
// manual_mode: This file has been manually modified and should not be touched by generate_services.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const grpc = __importStar(require("google-ads-node"));
const utils_1 = require("google-ads-node/build/lib/utils");
const wrappers_pb_1 = require("google-protobuf/google/protobuf/wrappers_pb");
const service_1 = __importDefault(require("./service"));
class CustomerService extends service_1.default {
    constructor(cid, client, throttler, name, pre_report_hook, post_report_hook) {
        super(cid, client, throttler, name);
        this.pre_report_hook = pre_report_hook;
        this.post_report_hook = post_report_hook;
    }
    report(options) {
        return __awaiter(this, void 0, void 0, function* () {
            const results = yield this.serviceReport(options, this.pre_report_hook, this.post_report_hook);
            return results;
        });
    }
    reportStream(options) {
        return this.serviceReportStream(options);
    }
    query(qry, options) {
        return __awaiter(this, void 0, void 0, function* () {
            const results = yield this.serviceQuery(qry, options);
            return results;
        });
    }
    list() {
        return __awaiter(this, void 0, void 0, function* () {
            return this.getListResults('customer');
        });
    }
    get(id) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.serviceGet({
                request: 'GetCustomerRequest',
                resource: `customers/${id}`,
                method: 'getCustomer',
                entity_id: id,
            });
        });
    }
    update(customer, options) {
        return __awaiter(this, void 0, void 0, function* () {
            const request = new grpc.MutateCustomerRequest();
            const operation = new grpc.CustomerOperation();
            const pb = this.buildResource('Customer', customer);
            operation.setUpdate(pb);
            const mask = utils_1.getFieldMask(customer);
            operation.setUpdateMask(mask);
            request.setCustomerId(this.cid);
            request.setOperation(operation);
            if (options && options.hasOwnProperty('validate_only')) {
                request.setValidateOnly(options.validate_only);
            }
            yield new Promise((resolve, reject) => {
                this.service.mutateCustomer(request, (err, res) => {
                    if (err) {
                        reject(err);
                    }
                    else {
                        resolve(res);
                    }
                });
            });
        });
    }
    mutateResources(operations, options) {
        return __awaiter(this, void 0, void 0, function* () {
            const request = new grpc.MutateGoogleAdsRequest();
            request.setCustomerId(this.cid);
            if (options && options.hasOwnProperty('validate_only')) {
                request.setValidateOnly(options.validate_only);
            }
            if (options && options.hasOwnProperty('partial_failure')) {
                request.setPartialFailure(options.partial_failure);
            }
            const ops = [];
            for (const operation of operations) {
                if (!operation.hasOwnProperty('_resource')) {
                    throw new Error(`Missing "_resource" key on entity`);
                }
                const operation_resource_name = operation._resource;
                let operation_mode = operation._operation;
                delete operation._resource;
                delete operation._operation;
                /* Create the resource e.g. "CampaignBudget" */
                const pb = this.buildResource(operation_resource_name, operation);
                /* Create create|update operation of resource type e.g. "CampaignBudgetOperation" */
                // @ts-ignore Types are no use here
                const resource_operation = new grpc[`${operation_resource_name}Operation`]();
                if (!operation_mode) {
                    operation_mode = 'create';
                }
                if (operation_mode !== 'create' && operation_mode !== 'update' && operation_mode !== 'delete') {
                    throw new Error(`"_operation" field must be one of "create"|"update"|"delete"`);
                }
                if (operation_mode === 'create') {
                    resource_operation.setCreate(pb);
                }
                if (operation_mode === 'update') {
                    resource_operation.setUpdate(pb);
                    const update_mask = utils_1.getFieldMask(operation);
                    resource_operation.setUpdateMask(update_mask);
                }
                if (operation_mode === 'delete') {
                    // @ts-ignore Types are no use here
                    if (!pb.toObject().hasOwnProperty('resourceName') || !pb.toObject().resourceName) {
                        throw new Error(`Must specify "resource_name" to remove when using "delete"`);
                    }
                    // @ts-ignore Types are no use here
                    resource_operation.setRemove(pb.toObject().resourceName);
                }
                /* Add operation of resource type to global mutate operation e.g. "MutateOperation.setCampaignBudgetOperation" */
                const op = new grpc.MutateOperation();
                const operation_set_method = `set${operation_resource_name}Operation`;
                // @ts-ignore Types are no use here
                op[operation_set_method](resource_operation);
                /* Push operation to MutateOperationsList */
                ops.push(op);
            }
            request.setMutateOperationsList(ops);
            const response = yield this.globalMutate(request);
            return response;
        });
    }
    /**
     * Create new Customer/Account under Master Account
     * @ref https://developers.google.com/google-ads/api/reference/rpc/v3/CustomerService method CreateCustomerClient
     * @param {CreateCustomerOptions} options
     * @param {CreateCustomerFlowSettings} [flow_settings]
     * @param {boolean} [flow_settings.return_customer] Optional flag set to False by default. Provide True if you
     *   prefer to have Customer instance as result instead of CreateCustomerClientResponse
     */
    createCustomerClient(options, flow_settings = { return_customer: false }) {
        return __awaiter(this, void 0, void 0, function* () {
            if (flow_settings && flow_settings.return_customer && !flow_settings.gads_api) {
                throw new TypeError(`Missing 'gads_api' in 'flow_settings'.`);
            }
            const request = new grpc.CreateCustomerClientRequest();
            const customerClientPB = this.buildResource('Customer', options.customer_client);
            request.setCustomerId(options.customer_id);
            request.setCustomerClient(customerClientPB);
            if (options.access_role) {
                request.setAccessRole(options.access_role);
            }
            if (options.email_address) {
                const emailValue = new wrappers_pb_1.StringValue();
                emailValue.setValue(options.email_address);
                request.setEmailAddress(emailValue);
            }
            const response = (yield this.serviceCall('createCustomerClient', request));
            if (!flow_settings || !flow_settings.return_customer) {
                return response;
            }
            const customer_id = response.resource_name.split('/')[1];
            return flow_settings.gads_api.Customer({
                customer_account_id: customer_id,
                refresh_token: this.client.getRefreshToken(),
                login_customer_id: options.customer_id,
            });
        });
    }
}
exports.default = CustomerService;
