// google-ads-api/build/services/keyword_plan.js
"use strict";
// manual_mode: This file has been manually modified and should not be touched by generate_services.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const google_ads_node_1 = require("google-ads-node");
const service_1 = __importDefault(require("./service"));
/**
 * @constants
 */
const RESOURCE_URL_NAME = 'keywordPlans';
const MUTATE_METHOD = 'mutateKeywordPlans';
const MUTATE_REQUEST = 'MutateKeywordPlansRequest';
const OPERATION_REQUEST = 'KeywordPlanOperation';
const GET_METHOD = 'getKeywordPlan';
const GET_REQUEST = 'GetKeywordPlanRequest';
const RESOURCE = 'KeywordPlan';
class KeywordPlanService extends service_1.default {
    get(id) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.serviceGet({
                request: GET_REQUEST,
                resource: `${RESOURCE_URL_NAME}/${id}`,
                method: GET_METHOD,
                entity_id: id,
            });
        });
    }
    list(options) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.getListResults('keyword_plan', options);
        });
    }
    create(keyword_plan, options) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.serviceCreate(Object.assign({ request: MUTATE_REQUEST, operation: OPERATION_REQUEST, mutate: MUTATE_METHOD, entity: [RESOURCE, keyword_plan] }, options));
        });
    }
    update(keyword_plan, options) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.serviceUpdate(Object.assign({ request: MUTATE_REQUEST, operation: OPERATION_REQUEST, mutate: MUTATE_METHOD, entity: [RESOURCE, keyword_plan] }, options));
        });
    }
    delete(id, options) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.serviceDelete(Object.assign({ request: MUTATE_REQUEST, operation: OPERATION_REQUEST, mutate: MUTATE_METHOD, resource: `${RESOURCE_URL_NAME}/${id}`, entity_id: id }, options));
        });
    }
    generateForecastMetrics(id) {
        return __awaiter(this, void 0, void 0, function* () {
            const request = new google_ads_node_1.GenerateForecastMetricsRequest();
            request.setKeywordPlan(`customers/${this.cid}/${RESOURCE_URL_NAME}/${id}`);
            const response = yield new Promise((resolve, reject) => {
                this.service.generateForecastMetrics(request, (err, res) => {
                    if (err) {
                        reject(err);
                    }
                    else {
                        resolve(res);
                    }
                });
            });
            const [parsed] = this.parseServiceResults([response]);
            return parsed;
        });
    }
    generateHistoricalMetrics(id) {
        return __awaiter(this, void 0, void 0, function* () {
            const request = new google_ads_node_1.GenerateHistoricalMetricsRequest();
            request.setKeywordPlan(`customers/${this.cid}/${RESOURCE_URL_NAME}/${id}`);
            const response = yield new Promise((resolve, reject) => {
                this.service.generateHistoricalMetrics(request, (err, res) => {
                    if (err) {
                        reject(err);
                    }
                    else {
                        resolve(res);
                    }
                });
            });
            const [parsed] = this.parseServiceResults([response]);
            return parsed;
        });
    }
}
exports.default = KeywordPlanService;
